internal/security/proxy/service.go

```diff
@@ -0,0 +1,423 @@
+			return err
+		}
+
+		routeParams := &KongRoute{
+			Paths: []string{"/" + name},
+			Name:  name,
+		}
+		err = s.initKongRoutes(routeParams, name)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = s.initAuthMethod(Configuration.KongAuth.Name, Configuration.KongAuth.TokenTTL)
+	if err != nil {
+		return err
+	}
+
+	err = s.initACL(Configuration.KongACL.Name, Configuration.KongACL.WhiteList)
+	if err != nil {
+		return err
+	}
+
+	LoggingClient.Info("finishing initialization for reverse proxy")
+	return nil
+}
+
+func (s *Service) postCert(cert CertificateLoader) error {
+	cp, err := cert.Load()
+
+	if err != nil {
+		return err
+	}
+
+	body := &CertInfo{
+		Cert: cp.Cert,
+		Key:  cp.Key,
+		Snis: Configuration.SecretService.SNIS,
+	}
+	LoggingClient.Debug("trying to upload cert to proxy server")
+	data, err := json.Marshal(body)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return err
+	}
+	tokens := []string{Configuration.KongURL.GetProxyBaseURL(), CertificatesPath}
+	req, err := http.NewRequest(http.MethodPost, strings.Join(tokens, "/"), strings.NewReader(string(data)))
+	if err != nil {
+		LoggingClient.Error("failed to create upload cert request -- %s", err.Error())
+		return err
+	}
+	resp, err := s.client.Do(req)
+	if err != nil {
+		LoggingClient.Error("failed to upload cert to proxy server with error %s", err.Error())
+		return err
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusConflict:
+		LoggingClient.Info("successfully added certificate to the reverse proxy")
+		break
+	default:
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		e := fmt.Sprintf("failed to add certificate with errorcode %d, error %s", resp.StatusCode, string(b))
+		LoggingClient.Error(e)
+		return errors.New(e)
+	}
+	return nil
+}
+
+func (s *Service) initKongService(service *KongService) error {
+	formVals := url.Values{
+		"name":     {service.Name},
+		"host":     {service.Host},
+		"port":     {strconv.Itoa(service.Port)},
+		"protocol": {service.Protocol},
+	}
+	tokens := []string{Configuration.KongURL.GetProxyBaseURL(), ServicesPath}
+	req, err := http.NewRequest(http.MethodPost, strings.Join(tokens, "/"), strings.NewReader(formVals.Encode()))
+	if err != nil {
+		return fmt.Errorf("failed to construct http POST form request: %s %s", service.Name, err.Error())
+	}
+	req.Header.Add(clients.ContentType, "application/x-www-form-urlencoded")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		e := fmt.Sprintf("failed to set up proxy service: %s %s", service.Name, err.Error())
+		LoggingClient.Error(e)
+		return errors.New(e)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated:
+		LoggingClient.Info(fmt.Sprintf("successful to set up proxy service for %s", service.Name))
+		break
+	case http.StatusConflict:
+		LoggingClient.Info(fmt.Sprintf("proxy service for %s has been set up", service.Name))
```

Comment:
minor nitpick wording change
```suggestion
		LoggingClient.Info(fmt.Sprintf("proxy service for %s has already been set up", service.Name))
```